File: daily_python/2762_Continuous_Subarrays.py
import os
import sys
from typing import List

class Solution:
    def continuousSubarrays(self, nums: List[int]) -> int:
        freq = {}
        ans = 0
        start = 0
        for end in range(len(nums)):
            freq[nums[end]] = freq.get(nums[end], 0) + 1
            while max(freq) - 2 > min(freq):
                freq[nums[start]] -= 1
                if freq[nums[start]] == 0:
                    del freq[nums[start]]
                start += 1
            ans += end - start + 1
        return ans

def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            nums = list(map(int, lines[0].split('=')[1].strip()[1:-2].split(',')))
            dataset.append(nums)
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for nums in dataset:
        results.append(solution.continuousSubarrays(nums))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')
    

File: daily/2024/2762_Continuous_Subarrays.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    int count_neighbor(unordered_map<int,int>& m, int num){
        return m[num-2] + m[num-1] + m[num] + m[num+1] + m[num+2];
    }
    long long continuousSubarrays(vector<int>& nums) {
        long long ans = 0;
        unordered_map<int,int> m;
        int n = nums.size();
        int start = 0;
        for (int i = 0; i < n; i++){
            m[nums[i]]++;
            while(start < i && (i - start + 1) > count_neighbor(m, nums[i])){
                m[nums[start++]]--;
            }
            ans += (i - start + 1);
        }
        return ans;
    }
};

int main(){
    vector<int> nums = {5,4,2,4};
    // vector<int> nums = {1,2,3};
    Solution S;
    cout << S.continuousSubarrays(nums) << endl;
    return 0;
}